### PR TITLE
Register service worker in production

### PR DESCRIPTION
**Pull request by mrdrogdrog**
### Component/Part
Service worker

### Description
This PR activates the service worker for production builds.

Please review https://github.com/hedgedoc/react-client/pull/898 first!

### Steps

- [x] Added implementation
- [x] Extended changelog
- [x] I read the [contribution documentation](https://github.com/hedgedoc/react-client/blob/main/CONTRIBUTING.md) and signed-off my commits to accept the DCO.

### Related Issue(s)
#103

### CODE REVIEW

**Comment by DerMolly** [comment]
How does this get triggered? Currently it won't, or am I missing something?

**Comment by mrdrogdrog** [comment]
What do you mean?

**Review by DerMolly — APPROVED** [review]

**Comment by mrdrogdrog** [comment]
Okay now you can review this.

**Review by DerMolly — APPROVED** [review]

File: public/locales/en.json [hunk 1]
```diff
@@ -455,6 +455,15 @@
         "copyError": "Error while copying!",
         "errorOccurred": "An error occurred"
     },
+    "global": {
+        "modal": {
+            "reloadApp": {
+                "title": "Update available",
+                "content": "An update for hedgedoc is available. Please reload every hedgedoc tab.",
```
Comment (ErikMichelson) [inline public/locales/en.json:462]:
```suggestion
                "content": "An update for HedgeDoc is available. Please reload every HedgeDoc tab.",
```

File: src/service-worker-registration.ts [hunk 2]
```diff
@@ -115,36 +114,33 @@ function checkValidServiceWorker (swUrl: string, config?: Config) {
...
-      // Service worker found. Proceed as normal.
+        // Service worker found. Proceed as normal.
         registerValidSW(swUrl, config)
       }
     })
-    .catch(() => {
-      console.log('No internet connection found. App is running in offline mode.')
-    })
+  .catch(() => {
+    console.log('No internet connection found. App is running in offline mode.')
```
Comment (ErikMichelson) [inline src/service-worker-registration.ts:132]: Do we want to store a field `offline` in the redux (for example in the new notifications-object)? This way we could later show a "You're in offline-mode. Live-changes won't be applied."-message.

File: src/service-worker-registration.ts [hunk 3]
```diff
@@ -77,22 +76,22 @@ function registerValidSW (swUrl: string, config?: Config) {
...
-            // but the previous service worker will still serve the older
-            // content until all client tabs are closed.
+              // At this point, the updated precached content has been fetched,
+              // but the previous service worker will still serve the older
+              // content until all client tabs are closed.
               console.log(
                 'New content is available and will be used when all ' +
-              'tabs for this page are closed. See https://cra.link/PWA.'
+                'tabs for this page are closed. See https://cra.link/PWA.'
               )
```
Comment (ErikMichelson) [inline src/service-worker-registration.ts:85]: Customize or remove this message?
(For me it's not about the message but about the cra.link which could instead point to our FAQ)

File: src/service-worker-registration.ts [hunk 4]
```diff
@@ -32,37 +32,36 @@ type Config = {
...
-      }
-    })
-  }
+      // Add some additional logging to localhost, pointing developers to the
+      // service worker/PWA documentation.
+      navigator.serviceWorker.ready.then(() => {
+        console.log(
+          'This web app is being served cache-first by a service ' +
+          'worker. To learn more, visit https://cra.link/PWA'
+        )
```
Comment (ErikMichelson) [inline src/service-worker-registration.ts:58]: Same as with the other comment on cra.link I've already written.

File: public/locales/en.json [hunk 5]
```diff
@@ -455,6 +455,15 @@
...
+                "reloadButton": "Reload app"
```
Comment (ErikMichelson) [inline public/locales/en.json:463]: This in total is quite similar to another message we already have: `editor.error.newVersion`. To reduce the load on the translators, maybe we could use that instead or merge them?

**Review by ErikMichelson — CHANGES_REQUESTED** [review]
Although the service-worker itself works, the offline-support it should bring doesn't as the ApplicationInitializer fails to fetch the content from online-server. We should therefore probably cache the api/config as well.

File: src/service-worker-registration.ts (hunk 3 above)
Comment (mrdrogdrog) [reply src/service-worker-registration.ts:85]: I removed this message, because we only need the callback.

File: public/locales/en.json (hunk 5 above)
Comment (mrdrogdrog) [reply public/locales/en.json:463]: I would remove it becaus its no error and we have a new modal for that.

File: public/locales/en.json (hunk 1 above)
Comment (mrdrogdrog) [reply public/locales/en.json:462]: done

File: public/locales/en.json (hunk 5 above)
Comment (ErikMichelson) [reply public/locales/en.json]: Should we move the already translated strings to the new message if applicable?

Comment (mrdrogdrog) [reply public/locales/en.json]: It's another translation. And tbh I don't like the old translation.

**Comment by mrdrogdrog** [comment]
Maybe we should delete this branch and rewrite the service worker as soon as we know what we want to cache offline. Like API stuff

**Comment by ErikMichelson** [comment]
But why should we delete it then? Seems like we could extend it later.

**Pull request closed without merging**